core/commands/storage/retry.go

```diff
@@ -0,0 +1,116 @@
+package storage
+
+import (
+	"github.com/Workiva/go-datastructures/queue"
+	"sort"
+	"sync"
+	"time"
+)
+
+type RetryQueue struct {
+	Queue *queue.Queue
+}
+
+type HostNode struct {
+	sync.Mutex
+
+	Identity   string
+	Score      int64
+	RetryTimes int
+	FailTimes  int
+}
+
+func (node *HostNode) IncrementRetry() {
+	node.Lock()
+	defer node.Unlock()
+
+	node.RetryTimes++
+}
+
+func (node *HostNode) GetRetryTimes() int {
+	node.Lock()
+	defer node.Unlock()
+
+	return node.RetryTimes
+}
+
+func (node *HostNode) IncrementFail() {
+	node.Lock()
+	defer node.Unlock()
+
+	node.FailTimes++
+}
+
+func (node *HostNode) GetFailTimes() int {
+	node.Lock()
+	defer node.Unlock()
+
+	return node.FailTimes
+}
+
+type Hosts []*HostNode
+
+func (h Hosts) Len() int {
+	return len(h)
+}
+
+func (h Hosts) Less(i, j int) bool {
+	return h[i].Score > h[j].Score
+}
+
+func (h Hosts) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *Hosts) Add(node *HostNode) {
+	*h = append(*h, node)
+}
+
+func (h Hosts) Sort() {
+	sort.Sort(h)
+}
+
+// init retry queue with 30 initial capacity
+func New(capacity int64) *RetryQueue {
+	r := &RetryQueue{Queue: queue.New(capacity)}
+	return r
+}
+
+func (q *RetryQueue) AddAll(h Hosts) error {
```

Comment:
Can you add at least one unit test in `retry_test.go` covering all `RetryQueue` related functions? or one test for each function, depending on how you setup the test

Reply:
I am not actually using this function `AddAll` after i merge the code with weiyu's. I can delete it, but I save it for future use, when a time we need to define the priority of diverse source of hosts.

Currently I am using third-party queue's add, and all other functions, the only difference is that I cast interface to my struct type. The library has been tested with thread-safe, so I would say it's ok. But I agree we should have unit test on all those codes including command, not simply this one. However, for now, I would suggest it should stay low priority and a backlog ticket for it.

Reply:
Do queue functions first and others later. Unit testing should be a must have going forward (test coverage does not need to be high, only the core functions).

As for commands, we will have integration tests on those, unit tests would be too complex. Unit tests apply to independent routines and/or utils (such as a RetryQueue).